internal/cloudsql/refresh_test.go

```diff
@@ -15,77 +15,334 @@
+			wantErr: "failed to get instance",
+			desc:    "When the Metadata call fails",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(cn.project, cn.region, cn.name,
+					mock.WithRegion("some-other-region")), 1),
+			wantErr: "region was mismatched",
+			desc:    "When the region does not match",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(
+					cn.project, cn.region, cn.name,
+					mock.WithRegion("my-region"),
+					mock.WithFirstGenBackend(),
+				), 1),
+			wantErr: "only Second Generation",
+			desc:    "When the instance isn't Second generation",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(
+					cn.project, cn.region, cn.name,
+					mock.WithRegion("my-region"),
+					mock.WithMissingIPAddrs(),
+				), 1),
+			wantErr: "no supported IP addresses",
+			desc:    "When the instance has no supported IP addresses",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(
+					cn.project, cn.region, cn.name,
+					mock.WithRegion("my-region"),
+					mock.WithSigner(func(_ *x509.Certificate, _ *rsa.PrivateKey) ([]byte, error) {
+						return nil, nil
+					}),
+				), 1),
+			wantErr: "failed to decode",
+			desc:    "When the server cert does not decode",
+		},
+		{
+			req: mock.InstanceGetSuccess(
+				mock.NewFakeCSQLInstance(
+					cn.project, cn.region, cn.name,
+					mock.WithRegion("my-region"),
+					mock.WithSigner(func(_ *x509.Certificate, _ *rsa.PrivateKey) ([]byte, error) {
+						certPEM := &bytes.Buffer{}
+						pem.Encode(certPEM, &pem.Block{
+							Type:  "CERTIFICATE",
+							Bytes: []byte("hello"), // woops no cert
+						})
+						return certPEM.Bytes(), nil
+					}),
+				), 1),
+			wantErr: "failed to parse",
+			desc:    "When the cert is not a valid X.509 cert",
+		},
+	}
+	for i, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			mc, url, cleanup := mock.HTTPClient(tc.req)
+			client, err := sqladmin.NewService(
+				context.Background(),
+				option.WithHTTPClient(mc),
+				option.WithEndpoint(url),
+			)
+			if err != nil {
+				t.Fatalf("failed to create test SQL admin service: %s", err)
+			}
+			defer cleanup()
+
+			r := newRefresher(time.Hour, 30*time.Second, 1, client)
+			_, _, _, err = r.performRefresh(context.Background(), cn, RSAKey)
+
+			if !errorContains(err, tc.wantErr) {
+				t.Errorf("[%v] PerformRefresh failed with unexpected error, want = %v, got = %v", i, tc.wantErr, err)
+			}
+		})
+	}
+}
+
+func TestRefreshWithFailedEphemeralCertCall(t *testing.T) {
+	cn, _ := parseConnName("my-project:my-region:my-instance")
 	inst := mock.NewFakeCSQLInstance(cn.project, cn.region, cn.name)
 
-	// mock expected requests
+	testCases := []struct {
+		req     *mock.Request
+		wantErr string
+		desc    string
+	}{
+		{
+			req:     mock.InstanceGetSuccess(inst, 1), // not an ephemeral cert call
+			wantErr: "create failed",
+			desc:    "When the CreateEphemeralCert call fails",
+		},
```

Comment:
There are three additional test cases to add here, but they'll require registering a separate instance with the mock server such that the metadata call succeeds but the fetch cert call fails (i.e., 1. when MarshalPKIXPublicKey fails, 2. When the cert response fails to decode, and 3. When the X.509 cert does not parse).

Reply:
Can you change req to a slice of requests instead and include those tests here?

Reply:
Will do.